rest-client/src/main/java/org/sagebionetworks/bridge/rest/AuthenticationHandler.java

```diff
@@ -1,96 +1,73 @@
 package org.sagebionetworks.bridge.rest;
 
-import java.io.IOException;
+import static com.google.common.base.Preconditions.checkNotNull;
 
-import com.google.common.base.Strings;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import java.io.IOException;
 
-import org.sagebionetworks.bridge.rest.model.SignIn;
 import org.sagebionetworks.bridge.rest.model.UserSessionInfo;
 
-import com.google.common.base.Preconditions;
-
 import okhttp3.Authenticator;
 import okhttp3.Interceptor;
+import okhttp3.Request;
+import okhttp3.Response;
 import okhttp3.Route;
 
 /**
- * Authenticates a user with Bridge: retrieves session and attaches token to header.
+ * Intercepts authentication error responses and re-acquires a session. NOTE: This handler is not currently registered
+ * by the ApiClientProvider to handle authentication, because no integration test indicates that it is needed. When we
+ * detect a session is missing, we reauthenticate using the UserSessionInfoProvider. It may be that this is needed when
+ * a session times out on the server, if so, no test currently simulates this behavior.
  */
-class AuthenticationHandler implements Authenticator, Interceptor {
-    private static final Logger LOG = LoggerFactory.getLogger(AuthenticationHandler.class);
-    private static final int MAX_TRIES = 3;
+class AuthenticationHandler implements Interceptor, Authenticator {
 
-    private final SignIn signIn;
+    private static final String SIGN_OUT_PATH = "/signOut";
+    private static final String AUTH_PATH = "/auth/";
+    
     private final UserSessionInfoProvider userSessionInfoProvider;
 
-    private UserSessionInfo userSession;
-    private int tryCount;
-
-    /**
-     * @param signIn
-     *         credentials for the user
-     * @param userSessionInfoProvider
-     *         for retrieving the user's session
-     */
-    public AuthenticationHandler(SignIn signIn, UserSessionInfoProvider userSessionInfoProvider) {
-        Preconditions.checkNotNull(signIn);
-        Preconditions.checkNotNull(userSessionInfoProvider);
-        this.signIn = signIn;
+    public AuthenticationHandler(UserSessionInfoProvider userSessionInfoProvider) {
+        checkNotNull(userSessionInfoProvider);
         this.userSessionInfoProvider = userSessionInfoProvider;
     }
-
-
+    
     @Override
-    public okhttp3.Request authenticate(Route route, okhttp3.Response response) throws IOException {
-        // if we reach this part of the code, the server had returned a 401 and userSession is
-        // invalid
-        userSessionInfoProvider.purgeCachedSession(this.userSession);
-        this.userSession = null;
-
-        if (tryCount >= MAX_TRIES || !isValidSignIn(signIn)) {
-            LOG.info("Maximum retries reached");
-            this.tryCount = 0;
-
-            // returning null terminates the request chain and propagates the 401
-            return null;
+    public Request authenticate(Route route, Response response) throws IOException {
```

Comment:
Without a retry count here, are we avoiding infinite loops for when we keep getting 401s (e.g. password changed).

Reply:
On second thought, I think checking if sessionToken is present after reauth, and return null if sessionToken=null will handle it. No retry needed.